### PR TITLE
ci(portal loop): Create exports from portal loop

### PR DESCRIPTION
**Pull request by sw360cab**
Introduces a CI workflow that backup deltas of new transactions from portal loop and perform the required extractions from there

### CODE REVIEW

File: .github/workflows/portal-loop-txs-exporter.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+name: Backup Portal Loop
+
+on:
+  # allow to run workflow manually
+  workflow_dispatch: { }
+
+  # Triggers the workflow every day at 1am
```
Comment (zivkovicmilos) [inline .github/workflows/portal-loop-txs-exporter.yml:7]: Isn't this every hour?

**Review by zivkovicmilos — APPROVED** [review]

**Review by moul — COMMENTED** [review]
can you create a new PR with only the logic and without the result? (feel free to keep this one open as a way to preview the result).

**Comment by moul** [comment]
I don't believe a CI workflow is the best approach. The portal loop isn't a stable network, so you can easily miss transactions by not calling this workflow at the right moment, which will likely happen about 99.9% of the time.

The portal loop already has a concept of local logs that are created on the machine precisely when needed, specifically just before an upgrade.

This is likely what is already backed up on S3, which isn't what we want. We need a way to patch the logs, which is why we require this repository instead of s3.

File: .github/workflows/portal-loop-txs-exporter.yml (hunk 1 above)
Comment (sw360cab) [reply .github/workflows/portal-loop-txs-exporter.yml:7]: yes it is every hour, forgot to update the comment

**Comment by sw360cab** [comment]
I can do the PR, I just lost track of this since it is not clear to me what is the final expected outcome.

The way it was thought for the portal loop is that we are saving only the _deltas_ from last / most recent backup.
So there will a very first save of the whole transactions, following stored files (monotonic), will only store the deltas from the previous backup. 
In this sense:

- using timestamps and sort guarantees non overlapping data
- the moment of saving is not so important since the delta stored will go back to the latest save (sort + diff)

A way to make this more effective is to connect the job to the merge into `master` of the monorepo so that the save occur before portal loop is looped.
Or we should embed all this stuff into the "looping" process of the portal-loop itself. To me the latter is way more complicated.

**Comment by moul** [comment]
Can you do the PR, please?

**Comment by sw360cab** [comment]
The PR has been updated leaving only the relevant script (used into the workflow).

We reviewed the flow to make it simple and clear:
* there is only one backup reference
  * a file for txs
  * a file for balances
* the script checks every hour the status of PL
* the first time the file will just save the backup files as nothing else was found
* following tasks dispatched (cron) will compare the current txs export with the prevoius one
* if no differences are found nothing happens
* otherwise the incoming exported txs file (and separeted balances file) will replace existing one

Since any updated file should triggers a commit in the repo, we should actually achieve out of-the-box history and versioning of the backup reference.

File: portal-loop/export.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+# Install the gnoland binary
+git clone https://github.com/gnolang/gno.git
+cd gno/gno.land || exit 1
+make build.gnoland
+cd ../.. # move back to the portal-loop directory
+
+# Extract the genesis transactions
+./gno/gno.land/build/gnoland genesis txs export -genesis-path $GENESIS "$BACKUP_NAME_TXS"
+# Extract the genesis transactions
```
Comment (zivkovicmilos) [inline portal-loop/export.sh:38]: Typo here, `genesis balances`

File: portal-loop/export.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# Define the constants
+TMP_DIR=temp-gno # Temporary directory for the working data
+GENESIS=genesis.json
+WGET_OUTPUT=wget-genesis.json
+BACKUP_NAME_TXS=backup_portal_loop_txs.jsonl
+BACKUP_NAME_BALANCES=backup_portal_loop_balances.jsonl
+# Latest backup is commited in the repo
+LATEST_BACKUP_FILE_TXS=../"$BACKUP_NAME_TXS"
+
+# Make the generated backup files the reference ones stored into the repositor
```
Comment (zivkovicmilos) [inline portal-loop/export.sh:14]: Typo here `repository`

File: portal-loop/export.sh [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  # There is an existing backup already, check it
+  echo "Backup file already present in the repository"
+
+  # Sort the latest backup file
+  sort "$LATEST_BACKUP_FILE_TXS" > "$LATEST_BACKUP_SORTED"
+  
+  # Sort the latest genesis tx sheet
+  sort "$BACKUP_NAME_TXS" > temp_"$BACKUP_NAME_TXS"
+
+  # Use comm to find lines only in file2 (additions) and write to output file
```
Comment (zivkovicmilos) [inline portal-loop/export.sh:64]: We should fixup this comment

File: portal-loop/export.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  comm -13 ./"$LATEST_BACKUP_SORTED" temp_"$BACKUP_NAME_TXS" > "$DIFF_TXS"
+
+  # Notify if differences were found
+  if [[ -z $(grep '[^[:space:]]' "$DIFF_TXS") ]]; then
+    echo "No differences found. Exiting with no further activities."
+  else
+    echo "Differences found. Replacing backup files"
+    # Make the backup files official
+    copyBackupFiles
+    echo "Stored new backup files for balances and txs"
```
Comment (zivkovicmilos) [inline portal-loop/export.sh:74]: Just a note, it's fine to replace both the balances and txs, because balances can't be changed without transactions

**Review by zivkovicmilos — APPROVED** [review]
Looks good 😎 

Pinging @moul for a final review 🙏

**Comment by zivkovicmilos** [comment]
@moul 

The idea behind this approach is to always have a git history of the portal loop transactions (that are equal to the ones on the machine), so we can change the history, and easily "replay" the portal loop with the modified genesis balances / txs

**Review by moul — APPROVED** [review]

**Pull request merged**